### ISSUE TITLE
Feedback on the Compiler Documentation Guides

### ISSUE THREAD

**Issue by Dominik1999** [0xPolygonMiden/compiler#104]
I have done the compiler documentation guides:

- https://0xpolygonmiden.github.io/compiler/guides/rust_to_wasm.html
- https://0xpolygonmiden.github.io/compiler/guides/wasm_to_masm.html

Here is a quick feedback.

## Rust to WASM
Nicely written and easy to follow. 

### rust-toolchain.toml
Nit, I needed to put `"wasm32-unknown-unknown"` into brackets to make it work (I believe `targets` expects an array of values? So my `rust-toolchain.toml` looked like:

```
[toolchain]
channel = "nightly"
targets = ["wasm32-unknown-unknown"]
```

### src/lib.rs
works fine and I like the well written comments to explain everything. 

Only `panic()` works without `unsafe` (at least my compiler complained)

```
// Required for no-std crates
#[panic_handler]
fn panic(info: core::panic::PanicInfo) -> ! {
    core::intrinsics::abort()
}
```

### compiling to WebAssembly
there is some error in the docs?

<img width="1213" alt="Screenshot 2024-01-25 at 12 20 09" src="https://github.com/0xPolygonMiden/compiler/assets/35031754/19b858b9-494a-477d-81c3-eb417298465e">

It seems not to be shown correctly. At least the GH page doesn't render what is here https://github.com/0xPolygonMiden/compiler/blob/main/docs/src/guides/rust_to_wasm.md

### Remove the $ from the code snippets
Tiny nit: can you remove the $ from the code snippets, then I can copy and paste into my editor. 

```
$ cargo build --release --target=bla
```

-> 

```
cargo build --release --target=bla
```

## WASM to Masm
Nicely written and easy to follow. Two minor points:

1. Somehow the docs that are built [here](https://0xpolygonmiden.github.io/compiler/guides/wasm_to_masm.html) differ from `main` ([here](https://github.com/0xPolygonMiden/compiler/blob/main/docs/src/guides/wasm_to_masm.md)). Is there an old version built?

2. Can you explain the example a bit? Where does `push.4294967295` come from? 

3. Would be cool to use the compiled Masm file directly in the playground. For that, we should have a program (starting with `begin`) rather than a procedure (? starting with export). Then ppl could try the compiler example without needing to get the VM as well.

**Comment by bitwalker** [comment]
Thanks for the thorough review @Dominik1999! I'll go through and fix those typos and things today.

> Somehow the docs that are built [here](https://0xpolygonmiden.github.io/compiler/guides/wasm_to_masm.html) differ from main ([here](https://github.com/0xPolygonMiden/compiler/blob/main/docs/src/guides/wasm_to_masm.md)). Is there an old version built?

The `main` branch hasn't been merged to in a long time, as we have not done any real releases per se. We can certainly start to do that on a regular basis though. I'll make a note to do that today as well.

> Can you explain the example a bit? Where does push.4294967295 come from?

Like, explain it in the docs, or do you mean in a more meta-documentation sense, such as why I chose that specific example/etc? The constants in the generated code are almost always bitmasks to handle things like signed arithmetic using unsigned u32 values, occasionally its for assertions related to type casts, things of that nature, but it depends on the specific instruction sequence. That constant looks to me like `u32::MAX`, which could be getting used for a variety of reasons, I'd have to look at the IR to know for sure. Happy to add a note to the docs about that if you think it's something people are likely to be curious about.

> Would be cool to use the compiled Masm file directly in the playground. For that, we should have a program (starting with begin) rather than a procedure (? starting with export). Then ppl could try the compiler example without needing to get the VM as well.

The example code here was being specifically compiled as a library to demonstrate a few other bits of the tooling, but we can emit programs as well, though in general what the compiler produces is probably not in an ideal form for the playground, since the assembler complains if we have "exported" procedures in a program module, and the compiler uses visibility information provided by the source language to determine whether a procedure is exported or not. Bottom line, we emit the program module (i.e. the module with `begin`) separately from everything else, and explicitly import things that are used. From what I gather, the playground is oriented more towards single-file examples. All very solvable things, but might preclude just compiling a program and then running it in the playground without any modification at the moment.

**Comment by bitwalker** [comment]
I believe all of your feedback has been addressed in #126, though there were a few questions above I had regarding one or two of the suggestions. I'm going to close this for now, but feel free to reopen if further action is required on any of those outstanding items. Thanks again for the thorough review!
